src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/RefDictionary.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Experimental.Collections
+{
+    public class RefDictionary<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey, TValue>> where TKey : IEquatable<TKey> // almost IReadOnlyDictionary<TKey, TValue>
```

Comment:
A better name for this maybe `DictionarySlim`. It is the naming convention we have used in other places for leaner equivalents of classic types, e.g. `Semapore`/`SemaphoreSlim`, `ReaderWriterLock`/`ReaderWriterLockSlim`, etc.

Reply:
Good call.

Reply:
This would not be a good collection for many reference types, whereas `ReaderWriterLockSlim` is (I think) always better than `ReaderWriterLock`. I do see `SemaphoreSlim` and `ManualResetEventSlim` do have more limited functionality than their non-slim counterparts, but I wonder whether `DictionarySlim` would appear to be the ideal choice if you don't need to supply a comparer, whereas it might have worse performance characteristics (eg., if your keys' GetHashCode or Equals implementations are slow)

Reply:
> This would not be a good collection for many reference type

This dictionary should work great for many reference types too. Whether it works great depends on the cost of the GetHashCode and Equals as you have said. It does not depend on whether the key type is a reference type.

> `ReaderWriterLockSlim` is (I think) always better than `ReaderWriterLock`

It is not always better. `ReaderWriterLock` is fair.  `ReaderWriterLockSlim` is not fair. Folks who need a fair lock cannot use `ReaderWriterLockSlim`.

Reply:
Alright. Pretty cool to have the prestigious name.